// @giscus/vue/dist/index.mjs
import { defineComponent as n, ref as l, onMounted as a, openBlock as i, createElementBlock as o, createCommentVNode as u } from "vue";
const d = ["id", "host", "repo", "repoid", "category", "categoryid", "mapping", "term", "strict", "reactionsenabled", "emitmetadata", "inputposition", "theme", "lang", "loading"], g = /* @__PURE__ */ n({
  __name: "Giscus",
  props: {
    id: null,
    host: null,
    repo: null,
    repoId: null,
    category: null,
    categoryId: null,
    mapping: null,
    term: null,
    theme: null,
    strict: null,
    reactionsEnabled: null,
    emitMetadata: null,
    inputPosition: null,
    lang: null,
    loading: null
  },
  setup(e) {
    const t = l(!1);
    return a(() => {
      t.value = !0, import("./giscus-a53917f0.mjs");
    }), (c, m) => t.value ? (i(), o("giscus-widget", {
      key: 0,
      id: e.id,
      host: e.host,
      repo: e.repo,
      repoid: e.repoId,
      category: e.category,
      categoryid: e.categoryId,
      mapping: e.mapping,
      term: e.term,
      strict: e.strict,
      reactionsenabled: e.reactionsEnabled,
      emitmetadata: e.emitMetadata,
      inputposition: e.inputPosition,
      theme: e.theme,
      lang: e.lang,
      loading: e.loading
    }, null, 8, d)) : u("", !0);
  }
});
export {
  g as default
};
