// @unocss/preset-mini@0.52.7/dist/rules.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const transform = require('./shared/preset-mini.c0eda31d.cjs');
const _default = require('./shared/preset-mini.b1ff13dc.cjs');
require('./shared/preset-mini.fcd935e5.cjs');
require('@unocss/core');



exports.alignments = transform.alignments;
exports.appearance = transform.appearance;
exports.appearances = transform.appearances;
exports.bgColors = transform.bgColors;
exports.borderStyles = transform.borderStyles;
exports.borders = transform.borders;
exports.boxShadows = transform.boxShadows;
exports.boxShadowsBase = transform.boxShadowsBase;
exports.boxSizing = transform.boxSizing;
exports.breaks = transform.breaks;
exports.contains = transform.contains;
exports.contentVisibility = transform.contentVisibility;
exports.contents = transform.contents;
exports.cursors = transform.cursors;
exports.displays = transform.displays;
exports.flexGridJustifiesAlignments = transform.flexGridJustifiesAlignments;
exports.floats = transform.floats;
exports.fontSmoothings = transform.fontSmoothings;
exports.fontStyles = transform.fontStyles;
exports.fonts = transform.fonts;
exports.handlerBorderStyle = transform.handlerBorderStyle;
exports.insets = transform.insets;
exports.justifies = transform.justifies;
exports.opacity = transform.opacity;
exports.orders = transform.orders;
exports.outline = transform.outline;
exports.overflows = transform.overflows;
exports.placements = transform.placements;
exports.pointerEvents = transform.pointerEvents;
exports.positions = transform.positions;
exports.resizes = transform.resizes;
exports.ringBase = transform.ringBase;
exports.rings = transform.rings;
exports.tabSizes = transform.tabSizes;
exports.textAligns = transform.textAligns;
exports.textColors = transform.textColors;
exports.textDecorations = transform.textDecorations;
exports.textIndents = transform.textIndents;
exports.textOverflows = transform.textOverflows;
exports.textShadows = transform.textShadows;
exports.textStrokes = transform.textStrokes;
exports.textTransforms = transform.textTransforms;
exports.textWraps = transform.textWraps;
exports.transformBase = transform.transformBase;
exports.transforms = transform.transforms;
exports.transitions = transform.transitions;
exports.userSelects = transform.userSelects;
exports.varEmpty = transform.varEmpty;
exports.verticalAligns = transform.verticalAligns;
exports.whitespaces = transform.whitespaces;
exports.willChange = transform.willChange;
exports.zIndexes = transform.zIndexes;
exports.aspectRatio = _default.aspectRatio;
exports.containerParent = _default.containerParent;
exports.cssProperty = _default.cssProperty;
exports.cssVariables = _default.cssVariables;
exports.flex = _default.flex;
exports.gaps = _default.gaps;
exports.grids = _default.grids;
exports.margins = _default.margins;
exports.paddings = _default.paddings;
exports.questionMark = _default.questionMark;
exports.rules = _default.rules;
exports.sizes = _default.sizes;
exports.svgUtilities = _default.svgUtilities;
